### PR TITLE
LODManager uses batchTime and engineRunTime for automatic LOD adjustment

### PR DESCRIPTION
**Pull request by AndrewMeadows**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3579/HighFidelity-Beta-PR11447-dfeee40a629fd534116b5322d1c770917f89bb7f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3579/HighFidelity-Beta-PR11447-dfeee40a629fd534116b5322d1c770917f89bb7f.exe

**Comment by AndrewMeadows** [comment]
The `LODManager` used to use the main loop rate for automatic LOD adjust, however it should use the `batchTime` (on the batch thread) and the render `engineRunTime` (on the present thread) for this purpose. In this PR the `LODManager` maintains a time-weighted running average of the max of these two times (with some minor offsets) and computes a "minimum angular diameter" for the `CullTask` render job to use for filtering out small stuff.

**Test Plan**

**(1)** Login to a domain that has a fair amount of objects in it, large and small.  dev-welcome should work

**(2)** Using Desktop mode (not HMD): navigate to a spot that has a range of object sizes in view: large to tiny

**(3)** Open the JS console (CTRL + ALT + j)

**(4)** Verify that the automatic LOD adjust feature is enabled.  This command should return `true`:
```
LODManager.getAutomaticLODAdjust();
```

**(5)** Turn off the automatic LOD adjust feature, and verify that it changed to 'false':
```
LODManager.setAutomaticLODAdjust(false);
LODManager.getAutomaticLODAdjust();
```

**(6)** Check the value of the minimum angular diameter:
```
LODManager.getMinAngularDiameter();
```
The default (minimum) value is just under 0.25 (degrees).

**(7)** Note the small and tiny objects that are in view and increase the minimum angular diameter to a degree or few:
```
LODManager.setMinAngularDiameter(2.0);
```
The smallest objects should disappear.

**(8)** Verify that the minimum angular diameter actually changed as reported by the LODManager:
```
LODManager.getMinAngularDiameter();
```

**(9)** Set the minimum angular diameter very small.  Verify that it gets small, but cannot be set lower than the default min (~ 0.25):
```
LODManager.setMinAngularDiameter(0.001);
LODManager.getMinAngularDiameter();
```

**(10)** Make the debug "stats" visible (using the / key, you need the "Developer menu" options enabled for this) and note the values in the upper right: "Engine" and "Batch".

**(11)** Change the minimum angular diameter to be large and verify that the "Engine" and "Batch" times decrease.  Raising the minimum angular diameter reduces the total number of renderable objects that reach the render pipeline, which reduces the time spent in these two contexts.

**(12)** Re-enable the automatic LOD adjustment:
```
LODManager.setAutomaticLODAdjust(true);
```

**(13)** If your video card/CPU is weak enough you may be able to see the automatic LOD adjustment kick in if you can find a sufficiently complex domain with many small objects.  When the render load gets too high the smaller objects should vanish as the LODManager tries to keep the render frame rate high.

**(14)** Finally, make sure the LOD auto adjust is working in HMD: if possible find a domain that has too many objects for your graphics card to handle in HMD mode and see if you can see the LOD changing (small objects vanishing or showing up again as the minimum angular diameter is automatically changed).

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3580/HighFidelity-Beta-PR11447-827788f411590cc0bc29c884b9c67c65888fe621.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3582/HighFidelity-Beta-PR11447-827788f411590cc0bc29c884b9c67c65888fe621.dmg

**Comment by AndrewMeadows** [comment]
Jenkins, rebuild this please.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -2428,8 +2428,12 @@ void Application::paintGL() {
             _viewFrustum.setProjection(adjustedProjection);
             _viewFrustum.calculate();
         }
-        renderArgs = RenderArgs(_gpuContext, lodManager->getOctreeSizeScale(),
-            lodManager->getBoundaryLevelAdjust(), RenderArgs::DEFAULT_RENDER_MODE,
+        // HACK: convert LODManager's state into a form that RenderArgs can understand
+        float octreeSizeScale = DEFAULT_OCTREE_SIZE_SCALE / lodManager->getLODScaleFactor();
+        int32_t boundaryLevelAdjust = 0;
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: why aren't you taking this value from the LOD manager?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3587/HighFidelity-Beta-PR11447-827788f411590cc0bc29c884b9c67c65888fe621.dmg

File: interface/src/LODManager.cpp [hunk 2]
```diff
@@ -19,223 +19,170 @@
...
+        if (!_automaticLODAdjust) {
+            _minAngularDiameter = ABSOLUTE_MIN_ANGULAR_DIAMETER;
+        }
+    }
 }
 
 void LODManager::loadSettings() {
-    setDesktopLODDecreaseFPS(desktopLODDecreaseFPS.get());
-    setHMDLODDecreaseFPS(hmdLODDecreaseFPS.get());
+    /* these Settings are for debug purposes: don't load
```
Comment (ZappoMan) [inline interface/src/LODManager.cpp]: why is this code commented out?

File: interface/src/LODManager.cpp [hunk 3]
```diff
@@ -19,223 +19,170 @@
...
+    /* these Settings are for debug purposes: don't save
+    automaticallyAdjustLOD.set(_automaticLODAdjust);
+    // save minAngularDiameter in degrees so that it is "human readable"
+    // and agrees with the units used in the UI
+    setMinAngularDiameter(glm::degrees(_minAngularDiameter));
+    */
 }
 
+void LODManager::setRenderMode(QString mode) {
+    // TODO: make mode FPS's configurable Settings rather than hard-coded defaults
```
Comment (ZappoMan) [inline interface/src/LODManager.cpp]: why did you remove the old code that had configurable amounts?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3587/HighFidelity-Beta-PR11447-827788f411590cc0bc29c884b9c67c65888fe621.exe

File: interface/src/ui/PreferencesDialog.cpp [hunk 4]
```diff
@@ -147,27 +147,6 @@ void setupPreferences() {
         auto setter = [](bool value) { Menu::getInstance()->setIsOptionChecked(MenuOption::DisableActivityLogger, !value); };
         preferences->addPreference(new CheckPreference("Privacy", "Send data", getter, setter));
     }
-    
-    static const QString LOD_TUNING("Level of Detail Tuning");
```
Comment (ZappoMan) [inline interface/src/ui/PreferencesDialog.cpp]: why did you remove this?

**Comment by Nex-Pro** [comment]
T E S T I N G this

**Comment by Nex-Pro** [comment]
### Result Undetermined as yet. Please read notes.


Test Plan

(1) Login to a domain that has a fair amount of objects in it, 
large and small. dev-welcome should work

(2) Using Desktop mode (not HMD): 
	navigate to a spot that has a range of object sizes in view: large to tiny

(3) Open the JS console (CTRL + ALT + j)

(4) Verify that the automatic LOD adjust feature is enabled. This command should return true:

LODManager.getAutomaticLODAdjust(); **SUCCESS**


(5) Turn off the automatic LOD adjust feature, and verify that it changed to 'false':

LODManager.setAutomaticLODAdjust(false); **DONE**
LODManager.getAutomaticLODAdjust(); **SUCCESS** false returned in console.


(6) Check the value of the minimum angular diameter:

LODManager.getMinAngularDiameter(); **DONE**
The default (minimum) value is just under 0.25 (degrees). **SUCCESS** Returned 0.2480980008840561



(7) Note the small and tiny objects that are in view and increase the minimum angular diameter to a degree or few:

LODManager.setMinAngularDiameter(2.0); **DONE**
The smallest objects should disappear. **SUCCESS** However there is an undesirable behaviour. For rotating items (Y axis) they tend to pop-in and pop-out as the object presents a variable horizontal angle across the bounding box. 



(8) Verify that the minimum angular diameter actually changed as reported by the LODManager:

LODManager.getMinAngularDiameter(); **See image...** Generally as expected but the image seems to indicate not always correct. Some time elapsed between setting at 2.0 and the reported 0.2480...

I N S E R T   I M A G E   H E R E

(9) Set the minimum angular diameter very small. Verify that it gets small, but cannot be set lower than the default min (~ 0.25):

LODManager.setMinAngularDiameter(0.001); **DONE**
LODManager.getMinAngularDiameter(); **SUCCESS** returns 0.2480980008840561

**CAUTION** Can set very large angles which are then returned. 

eg:

LODManager.setMinAngularDiameter(720);
LODManager.getMinAngularDiameter(); returns 720


(10) Make the debug "stats" visible (using the / key, you need the "Developer menu" options enabled for this) and note the values in the upper right: "Engine" and "Batch".

Engine ~3.5  Batch ~1.4


(11) Change the minimum angular diameter to be large and verify that the "Engine" and "Batch" times decrease.
Raising the minimum angular diameter reduces the total number of renderable objects that reach the render pipeline, which reduces the time spent in these two contexts.

**Screen went black while in focus when I set LODManager.setMinAngularDiameter(720); ** and Engine and Batch values dropped.
However when clicking on another app on the desktop the screen display returned and the Engine and Batch figures returned to previous levels.
Clicking back in interface the Engine and Batch figures dropped again.


(12) Re-enable the automatic LOD adjustment:

LODManager.setAutomaticLODAdjust(true); **DONE**



(13) If your video card/CPU is weak enough you may be able to see the automatic LOD adjustment kick in if you can find a sufficiently complex domain with many small objects. When the render load gets too high the smaller objects should vanish as the LODManager tries to keep the render frame rate high. Have gtx 1070 and i7.

(14) Finally, make sure the LOD auto adjust is working in HMD: if possible find a domain that has too many objects for your graphics card to handle in HMD mode and see if you can see the LOD changing (small objects vanishing or showing up again as the minimum angular diameter is automatically changed). **SEE IMAGE**

**Comment by Nex-Pro** [comment]
@AndrewMeadows I was in interface testing when you removed QA ready label

**Comment by AndrewMeadows** [comment]
Yeah sorry @Nex-Pro.  The work here was too ambitious. I'm scaling it back and trying again.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3611/HighFidelity-Beta-PR11447-cc8e618bc0379f3ec4dc8f8332e67b0531721aa7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3611/HighFidelity-Beta-PR11447-cc8e618bc0379f3ec4dc8f8332e67b0531721aa7.exe

**Comment by AndrewMeadows** [comment]
The changed behavior is that the LODManager now uses the "Engine" and "Batch" times (instead of main frame FPS) as input to when it needs to adjust LOD:

**New Test Plan:**

**(1)** Login to a domain with several objects (big and small).  Dev-welcome will probably work.

**(2)** Make sure your settings are the defaults by running this script:
```
LODManager.setAutomaticLODAdjust(true);
LODManager.setDesktopLODDecreaseFPS(20);
LODManager.setHMDLODDecreaseFPS(20);
```

**(3)** Note the values of the "Engine" and "Batch" times (upper right) in the stats (press '/' key).  Add two miliseconds to the avg "Batch" time and compare that sum with the avg value of "Engine".  Take the largest value (call it T) and compute maxFPS = 1/T.  This is the approximate max frame rate that the render pipeline could handle when rendering the current scene... if it weren't waiting fo the sync sync signal of your monitor (probably 60 Hz for desktop adn 90 Hz for HMD).

**(4)** Now set the "DesktopLODDecreaseFPS" setting to be something higher than your max available FPS.  For example, if you see 5msec then maxFPS would be 1/0.005 = 200 and you would set 300:
```
LODManager.setDesktopLODDecreaseFPS(300);
```
--> the LOD of the scene should chagne: small obects vanish and the time spent in the "Engine" and "Batch" should decrease.

**(5)** Run this command a few times as objects are vanishing:
```
LODManager.getOctreeSizeScale()
```
You may see a big number (2190081.5 or larger) move toward a smaller number (524288 or smaller).  What is the best way to understand this number?  Alas, I don't have a good explanation that can be very helpful, but let it be known that this number moves up and down with the LOD culling in proportion (or inversely, depending on how you think about it).

**(6)** While small objects are culled: open the "LOD Tools" window:  Developer --> Render --> LOD Tools.  Note the red text near the top: it says something about how your LOD vision compares with the average human eye.  If this makes sense then great.  If not: let it be known that the red text changes when the LOD culling changes.

**(7)** Set "DesktopLODDecreaseFPS" back to 20 and the small objects should show up again.  This could take several seconds for all the objects to show up.  The red text on the "LOD Tools" should update as the LOD culling changes.  The time spent in "Engine" and "Batch" in the stats should increase again.
```
LODManager.setDesktopLODDecreaseFPS(20);
```

**(8)** What happens when you run the above seven steps in hifi/master?  Objects will probably be culled (or not) and times spent in "Engine" and "Batch" will go down (or up) accordingly, but the automatic LOD feedback is not reliable since it is using the wrong stat (main loop FPS) --> all bets are off.

**Comment by AndrewMeadows** [comment]
@Nex-Pro, this is ready for test again, but the test has changed.

**Comment by Nex-Pro** [comment]
T e s t i n g   this

**Comment by Nex-Pro** [comment]
## PASS

New Test Plan:

(1) Login to a domain with several objects (big and small). Dev-welcome will probably work.

(2) Make sure your settings are the defaults by running this script:

LODManager.setAutomaticLODAdjust(true);
LODManager.setDesktopLODDecreaseFPS(20);
LODManager.setHMDLODDecreaseFPS(20);

(3) Note the values of the "Engine" and "Batch" times (upper right) in the stats (press '/' key). Add two miliseconds to the avg "Batch" time and compare that sum with the avg value of "Engine". Take the largest value (call it T) and compute maxFPS = 1/T. This is the approximate max frame rate that the render pipeline could handle when rendering the current scene... if it weren't waiting fo the sync sync signal of your monitor (probably 60 Hz for desktop adn 90 Hz for HMD).

(4) Now set the "DesktopLODDecreaseFPS" setting to be something higher than your max available FPS. For example, if you see 5msec then maxFPS would be 1/0.005 = 200 and you would set 300:

LODManager.setDesktopLODDecreaseFPS(300);
--> the LOD of the scene should change: small obects vanish and the time spent in the "Engine" and "Batch" should decrease. **SUCCESS**

(5) Run this command a few times as objects are vanishing:

LODManager.getOctreeSizeScale()
You may see a big number (2190081.5 or larger) move toward a smaller number (524288 or smaller). What is the best way to understand this number? Alas, I don't have a good explanation that can be very helpful, but let it be known that this number moves up and down with the LOD culling in proportion (or inversely, depending on how you think about it). **SUCCESS** Observed number  dropping from 7336194.5 to 524288 with LODManager.setDesktopLODDecreaseFPS(200);

(6) While small objects are culled: open the "LOD Tools" window: Developer --> Render --> LOD Tools. Note the red text near the top: it says something about how your LOD vision compares with the average human eye. If this makes sense then great. If not: let it be known that the red text changes when the LOD culling changes.

(7) Set "DesktopLODDecreaseFPS" back to 20 and the small objects should show up again. This could take several seconds for all the objects to show up. The red text on the "LOD Tools" should update as the LOD culling changes. The time spent in "Engine" and "Batch" in the stats should increase again. **SUCCESS**

LODManager.setDesktopLODDecreaseFPS(20);
(8) What happens when you run the above seven steps in hifi/master? Objects will probably be culled (or not) and times spent in "Engine" and "Batch" will go down (or up) accordingly, but the automatic LOD feedback is not reliable since it is using the wrong stat (main loop FPS) --> all bets are off. **Seemed to have little or no effect.**

(9) Make sure the manual LOD adjustment still works. Open the LODManager UI: Developer->Render->LOD_Tools. Check the "Manually Adjust Level Of Detail" box. Play with the slider moving it to the left (small objects should disappear) and to the right (small objects re-appear). Note: the red text with info about how the LOD compares to average human vision should update. When there is less LOD the stat time shown in "Engine" and "Batch" should decrease. **SUCCESS**
